### PR TITLE
improve summary log report (errors first + rm info lvl)

### PR DESCRIPTION
**Pull request by ftith**

### CODE REVIEW

File: nimp/summary.py [hunk 1]
```diff
@@ -240,25 +241,24 @@ def _add_notif(self, msg):
...
-                self._summary += '[  NOTIF  ] %s\n' % (self._context.popleft(),)
-        self._summary += '[  ERROR  ] %s\n' % (msg,)
+                self._errors_stack += ['[  NOTIF  ] %s\n' % (self._context.popleft(),)]
+        self._errors_stack += ['[  ERROR  ] %s\n' % (msg,)]
 
     def _write_summary(self, destination):
         ''' Writes summary to destination '''
-        destination.write(self._summary)
+        for lvl in ['errors', 'warnings']:
+            for line in self.__dict__["_{0}_stack".format(lvl)]:
```
Comment (jasugun) [inline nimp/summary.py:262]: if feels hack-ish to access `__dict__ `of an object directly.
We could access `_warnings_stack  `and `_errors_stack ` with getattr()?

```
for lvl in ['errors', 'warnings']:
            for line in getattr(self, f'_{lvl}_stack'):
```

Comment (jasugun) [reply nimp/summary.py:262]: Another options is to have `self._summary` be a dict of errors and warnings lists.

`self._summary['errors'] += ['[  ERROR  ] %s\n' % (msg,)]`
...etc 

then iterate the dict:
```
for lvl in ['errors', 'warnings']:
            for line in self._summary[lvl]:
```

**Review by jasugun — CHANGES_REQUESTED** [review]
I really feel acessing __dict__ is hacking and there are other options to achieve something similar.

File: nimp/summary.py [hunk 2]
```diff
@@ -49,6 +49,8 @@ def __init__(self, env):
         self._context_patterns = []
         self._has_errors = False
         self._has_warnings = False
```
Comment (tdesveaux) [inline nimp/summary.py:51]: Both `_has_errors` and `_has_warnings` could be replace with a check on the new `_stack` arrays

File: nimp/summary.py [hunk 3]
```diff
@@ -240,25 +241,24 @@ def _add_notif(self, msg):
...
+                self._errors_stack += ['[  NOTIF  ] %s\n' % (self._context.popleft(),)]
+        self._errors_stack += ['[  ERROR  ] %s\n' % (msg,)]
 
     def _write_summary(self, destination):
         ''' Writes summary to destination '''
-        destination.write(self._summary)
+        for lvl in ['errors', 'warnings']:
+            for line in self.__dict__["_{0}_stack".format(lvl)]:
+                destination.write(line)
+        destination.close()
```
Comment (tdesveaux) [inline nimp/summary.py:264]: `_write_summary` is called from a `with open` context, you do not need to close `destination` (and should not do this here. The place that open a file is responsible for closing it, else it leads to hard to follow code)

File: nimp/summary.py [hunk 4]
```diff
@@ -240,25 +241,24 @@ def _add_notif(self, msg):
...
-        self._summary += '[  ERROR  ] %s\n' % (msg,)
+                self._errors_stack += ['[  NOTIF  ] %s\n' % (self._context.popleft(),)]
+        self._errors_stack += ['[  ERROR  ] %s\n' % (msg,)]
 
     def _write_summary(self, destination):
         ''' Writes summary to destination '''
-        destination.write(self._summary)
+        for lvl in ['errors', 'warnings']:
+            for line in self.__dict__["_{0}_stack".format(lvl)]:
+                destination.write(line)
```
Comment (tdesveaux) [inline nimp/summary.py:263]: You could use `writeline` here and remove the line endings from the string you push in the stacks.

File: nimp/summary.py (hunk 1 above)
Comment (tdesveaux) [inline nimp/summary.py:262]: You could do something cleaner with:
```
for lvl in [self._errors_stack, self._warnings_stack]:
    destination.writelines(lvl)
```

File: nimp/summary.py [hunk 5]
```diff
@@ -240,25 +241,24 @@ def _add_notif(self, msg):
...
+        self._warnings_stack += ['[ WARNING ] %s\n' % (msg,)]
 
     def _add_error(self, msg):
+        self._errors_stack += ['\n *********************************************\n']
         if len(self._context) == self._context.maxlen:
-            self._summary += '\n *********************************************\n'
             while self._context:
-                self._summary += '[  NOTIF  ] %s\n' % (self._context.popleft(),)
-        self._summary += '[  ERROR  ] %s\n' % (msg,)
+                self._errors_stack += ['[  NOTIF  ] %s\n' % (self._context.popleft(),)]
```
Comment (tdesveaux) [inline nimp/summary.py:256]: a cleaner way would be:
```
if len(self._context) == self._context.maxlen:
    self._errors_stack += ['[ NOTIF ] %s\n' % (e,) for e in self._context]
    self._content.clear()
```

File: nimp/summary.py [hunk 6]
```diff
@@ -240,25 +241,24 @@ def _add_notif(self, msg):
...
+        self._warnings_stack += ['\n *********************************************\n']
         if len(self._context) == self._context.maxlen:
-            self._summary += '\n *********************************************\n'
             while self._context:
-                self._summary += '[  NOTIF  ] %s\n' % (self._context.popleft(),)
-        self._summary += '[ WARNING ] %s\n' % (msg,)
+                self._warnings_stack += ['[  NOTIF  ] %s\n' % (self._context.popleft(),)]
+        self._warnings_stack += ['[ WARNING ] %s\n' % (msg,)]
 
     def _add_error(self, msg):
```
Comment (tdesveaux) [inline nimp/summary.py:252]: Both `_add_error` and `_add_warning` could be refactored into one function that take 2 args:
- msg pre-formatted with `'[ WARNING ] %s\n' % (msg,)`
- stack (_errors_stack or warnings_stack)

File: nimp/summary.py (hunk 5 above)
Comment (ftith) [reply nimp/summary.py:256]: I've got the following error using self._content.clear()

`AttributeError: 'DefaultSummaryHandler' object has no attribute '_content'`

File: nimp/summary.py [hunk 7]
```diff
@@ -240,25 +230,15 @@ def _add_notif(self, msg):
...
-    def _add_warning(self, msg):
-        if len(self._context) == self._context.maxlen:
-            self._summary += '\n *********************************************\n'
-            while self._context:
-                self._summary += '[  NOTIF  ] %s\n' % (self._context.popleft(),)
-        self._summary += '[ WARNING ] %s\n' % (msg,)
 
-    def _add_error(self, msg):
+    def _add_msg(self, notif_lvl, msg):
+        self._summary['%ss' % notif_lvl] += ['\n *********************************************\n']
```
Comment (jasugun) [inline nimp/summary.py:235]: Definetely not the norm in nimp I know, but I think we should enforce fstrings whenever we can in our new code.
like `self._summary[f'{notif_lvl}s]`

File: nimp/summary.py [hunk 8]
```diff
@@ -167,11 +166,11 @@ def __exit__(self, ex_type, value, traceback):
 
     def has_errors(self):
         ''' Returns true if errors were emitted during program execution '''
-        return self._has_errors
+        return len(self._summary['errors']) > 0
```
Comment (tdesveaux) [inline nimp/summary.py:169]: Since you init with `{}`, key might not exists

**Review by tdesveaux — APPROVED** [review]

**Review by jasugun — APPROVED** [review]
Good job, simplified code, same level of functionnality ==  better
fstrings for the win :)

File: nimp/summary.py (hunk 8 above)
Comment (tdesveaux) [reply nimp/summary.py:169]: This initialization is done in `DefaultSummaryHandler` which is a subclass.
`has_` are in the parent (`SummaryHandler`) so if someone else inherit from it, this won't work.

File: nimp/summary.py [hunk 9]
```diff
@@ -240,25 +230,15 @@ def _add_notif(self, msg):
...
-    def _add_error(self, msg):
+    def _add_msg(self, notif_lvl, msg):
+        self._summary[f'{notif_lvl}s'] += ['\n *********************************************\n']
         if len(self._context) == self._context.maxlen:
-            self._summary += '\n *********************************************\n'
             while self._context:
-                self._summary += '[  NOTIF  ] %s\n' % (self._context.popleft(),)
-        self._summary += '[  ERROR  ] %s\n' % (msg,)
+                self._summary[f'{notif_lvl}s'] += ['[  NOTIF  ] %s\n' % (self._context.popleft(),)]
+        self._summary[f'{notif_lvl}s'] += ['[ ' + notif_lvl.upper() + ' ] %s\n' % (msg,)]
```
Comment (jasugun) [inline nimp/summary.py:239]: Doesn't the fstring work on the second part?  `self._summary[f'{notif_lvl}s'].append(f'[  {notif_lvl.upper()}  ] {msg}\n'])`

Comment (jasugun) [reply nimp/summary.py:239]: Usually, `append() ` and `extend()` function are more efficient that `+=` in python when handling lists.

**Review by jasugun — APPROVED** [review]
